src/metabase/models/pulse_channel.clj

```diff
@@ -0,0 +1,173 @@
+  [channel]
+  (boolean (:recipients? (get channel-types (keyword channel)))))
+
+
+(def days-of-week
+  "Simple `vector` of the days in the week used for reference and lookups.
+
+   NOTE: order is important here!!
+         these indexes match the values from clj-time `day-of-week` function (0 = Sunday, 6 = Saturday)"
+  [{:id "sun" :name "Sun"},
+   {:id "mon" :name "Mon"},
+   {:id "tue" :name "Tue"},
+   {:id "wed" :name "Wed"},
+   {:id "thu" :name "Thu"},
+   {:id "fri" :name "Fri"},
+   {:id "sat" :name "Sat"}])
+
+(def times-of-day
+  [{:id "morning" :name "Morning" :realhour 8},
+   {:id "midday" :name "Midday" :realhour 12},
+   {:id "afternoon" :name "Afternoon" :realhour 16},
+   {:id "evening" :name "Evening" :realhour 20},
+   {:id "midnight" :name "Midnight" :realhour 0}])
+
+(defn time-of-day->realhour
+  "Time-of-day to realhour"
+  [time-of-day]
+  (-> (filter #(= time-of-day (:id %)) times-of-day)
+      first
+      :realhour))
+
+
+;; ## Entity
+
+(defrecord PulseChannelInstance []
+  ;; preserve normal IFn behavior so things like ((sel :one Database) :id) work correctly
+  clojure.lang.IFn
+  (invoke [this k]
+    (get this k)))
+
+(extend-ICanReadWrite PulseChannelInstance :read :always, :write :superuser)
+
+(defentity PulseChannel
+  [(k/table :pulse_channel)
+   (hydration-keys pulse_channel)
+   (types :details :json, :schedule_details :json)
+   timestamped]
+
+  (post-select [_ {:keys [id creator_id details] :as pulse-channel}]
+    (map->PulseChannelInstance
+      (u/assoc* pulse-channel
+                ;; don't include `:emails`, we use that purely internally
+                :details    (dissoc details :emails)
+                ;; here we recombine user details w/ freeform emails
+                :recipients (delay (into (or (->> (:emails details)
+                                                  (mapv #(assoc {} :email %)))
+                                             [])
+                                         (db/sel :many [User :id :email :first_name :last_name]
+                                           (k/where {:id [in (k/subselect PulseChannelRecipient (k/fields :user_id) (k/where {:pulse_channel_id id}))]})))))))
+
+  (pre-cascade-delete [_ {:keys [id]}]
+    (db/cascade-delete PulseChannelRecipient :pulse_channel_id id)))
+
+(extend-ICanReadWrite PulseChannelEntity :read :always, :write :superuser)
+
+
+;; ## Helper Functions
+
+(defn update-recipients!
+  "Update the `PulseChannelRecipients` for PULSE-CHANNEL.
+   USER-IDS should be a definitive collection of *all* IDs of users who should receive the pulse.
+
+   *  If an ID in USER-IDS has no corresponding existing `PulseChannelRecipients` object, one will be created.
+   *  If an existing `PulseChannelRecipients` has no corresponding ID in USER-IDs, it will be deleted."
+  [id user-ids]
+  {:pre [(integer? id)
+         (coll? user-ids)
```

Comment:
not sure if this applies here but the predicate function `sequential?` can be used to ensure something is a sequential collection (i.e. a list or vector)